main/src/main/scala/sbt/Project.scala

```diff
@@ -121,7 +121,34 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
     if (ts.isEmpty) Nil else s"$label: $ts" :: Nil
 }
 
-sealed trait Project extends ProjectDefinition[ProjectReference] {
+trait CompositeProject {
+  def componentProjects: Seq[Project]
+}
+
+object CompositeProject {
+
+  /**
+   *  If two projects with the same id appear in `compositeProjects` and
+   *  in `compositeProjects.map(_.componentProjects)`, the one in
+   *  `compositeProjects` wins.
+   *  This is necessary for backward compatibility with the idiom:
+   *    lazy val foo = crossProject
+   *    lazy val fooJVM = foo.jvm.setting
+   */
+  def uniqueId(compositeProjects: Seq[Project]): Seq[Project] = {
```

Comment:
Basically, this method should take the original `Seq[CompositeProject]`, not a `Seq[Project]`. If it takes the `Seq[Project]`, it's too late: you don't know what were the projects that were in the initial `compositeProjects` versus those that were added through `componentProjects`.

Reply:
Hence, I would recommend calling this method `expand`, not `uniqueId`, since its job would be to *expand* the list of `CompositeProject` into the appropriate list of `Project`s.